presto-main/src/main/java/com/facebook/presto/execution/QueryLimit.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class QueryLimit<T extends Comparable<T>>
```

Comment:
Does the contract need to be so broad?  Anything that's comparable is a pretty broad contract.  Did we consider having separate classes for `Duration` and `DataSize`?

Reply:
I did think about having separate classes - but I didn't see any upside in having 2 different classes - it will just lead to code duplication. What do you think are the downsides of the broad contract ?

Reply:
Code duplication is fine if the alternative is less duplication but harder to reason about code.  To me, it's not immediately obvious what are the things that consist of the query limit when we have the generic bound above, and I needed to search through the code to understand and gain context that it's just `Duration` and `DataSize`.  It's a lot more straightforward to me that a query limit can come in the form of a duration or a size (things we measure queries on), hence why I think the duplication is worth it.

Reply:
Would you feel that a reasonable alternative might be to close the constructor (make it `private`) and have 2 constructor static methods - 

```
public static QueryLimit<Duration> createDurationLimit(Duration limit, Source source) {}
public static QueryLimit<DataSize> createDataSizeLimit(DataSize limit, Source source) {}
```

By doing so - it is obvious what all different limits are supported and there is no duplication of logic.

Reply:
Yeah, this would work.